### PR TITLE
New Plugin Spans

### PR DESCRIPTION
**Pull request by TheMutsi**
Expose Span methods to RN (recordSpanWithName is a custom method in JavaScript to manage the async functions)

Methods
```
startSpanWithName: (name: string, parentSpanId?: string) => Promise<boolean | string>;

stopSpanWithId: (spanId: string, errorCode?: SPAN_ERROR_CODES) => Promise<boolean>;

addSpanEventToSpanId: (spanId: string, name: string, timestamp?: number, attributes?: Attributes) => Promise<boolean>;

addSpanAttributesToSpanId: (spanId: string, key: string, value: string) => Promise<boolean>;

recordSpanWithName: (name: string, callback: () => void | Promise<void>, attributes?: Attributes, events?: Events[], parentSpanId?: string) => Promise<boolean>;

recordCompletedSpanWithName: (name: string, startTimestamp: number, endTimestamp: number, errorCode?: SPAN_ERROR_CODES, parentSpanId?: string, attributes?: Attributes, events?: Events[]) => Promise<boolean>;
```

Types
```
interface Attributes {
  [key: string]: string;
}
interface Events {
  name: string;
  timestamp?: number;
  attributes?: Attributes;
}

type SPAN_ERROR_CODES = 'None' | 'Failure' | 'UserAbandon' | 'Unknown';
```

Tests
Screenshot 2024-03-08 at 5 25 30 PM

Session Android (Spans + Breadcrumb + Network + Moments + Logs)
https://dash.embrace.io/app/p89ke/grouped_sessions/last_1_hour/F64F0004A27842C8AAB33E9A55DC3F91/2DED083F1C0B4EF9977B6D7AB54FE930@/2DED083F1C0B4EF9977B6D7AB54FE930?first=08B88F8B133344F6A4FA50D235AA33D4&last=2DED083F1C0B4EF9977B6D7AB54FE930

Session iOS (Spans + Breadcrumb + Network + Moments + Logs)
https://dash.embrace.io/app/A82U7/grouped_sessions/last_3_hour/17C1830B8BB443EE9B440703F2B5149C/56E0CD5FDB75479F9061FA6C68C75DCD@/56E0CD5FDB75479F9061FA6C68C75DCD?first=7B2E15C9490140D1B8215C43A66C9D75&last=56E0CD5FDB75479F9061FA6C68C75DCD

### CODE REVIEW

File: .github/CODEOWNERS
Comment (bidetofevil) [inline .github/CODEOWNERS:2]: It says this team is not public or doesn't have write access? Is it because you haven't created it yet?

File: .github/workflows/node.js.yml
Comment (bidetofevil) [inline .github/workflows/node.js.yml:18]: Android is using v4 now so it's probably safe for you to upgrade too

File: packages/action-tracker/README.md
Comment (bidetofevil) [inline packages/action-tracker/README.md:31]: This is just old code you're deleting?

File: packages/core/android/gradle.properties
Comment (bidetofevil) [inline packages/core/android/gradle.properties:1]: Whitespace change? Not sure sure if it's matters

File: packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java
Comment (bidetofevil) [inline packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java:620]: So Android should not hit this as Android returns a Long if that field is populated?

File: packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java
Comment (bidetofevil) [inline packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java:646]: So you're not adding to the spanId but to the span with the spanId.

Does it make sense to call this `addSpanAttributeToSpan()` or `addSpanAttributeWithSpanId()`?

Also it's singular - attribute not attributes 😅

File: packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java
Comment (bidetofevil) [inline packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java:655]: Same comment as above with the method name, re: ToSpan vs WithSpanId

File: packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java
Comment (bidetofevil) [inline packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java:665]: Lets use millis in the public API so we never get nanos coming from RN at all

File: packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java
Comment (bidetofevil) [inline packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java:669]: This method expects millis now starting with 6.4

File: packages/spans/src/index.ts
Comment (bidetofevil) [inline packages/spans/src/index.ts:48]: Do we usually add the unit (e.g. ms) to parameter names by convention? If it's ok, timeStampMs is preferred since the OTel JavaScript SDK uses nanos and we don't, so making that clear would be nice

This applies to all timestamps BTW.

File: packages/spans/src/index.ts
Comment (bidetofevil) [inline packages/spans/src/index.ts:11]: if you can do a default parameter to null here, that would be useful - most folks will be creating parent-less spans to start off with so not making them pass a null would be great.

or does TypeScript still do that for you if you drop trailing parameters still? JS did and I hated it so I was hoping TS would've removed that lol

**Review by bidetofevil — COMMENTED** [review]
I looked at the Java and JS code only (not iOS) and it mostly looks good.

The one thing that should change is the public API should expose milliseconds when it comes to timestamps (ideally it would be in the name of the parameter as well). Some other comments about method naming would be nice to do but it's up to you.

I'll approve it from my end once the time units are fixed for the API

File: packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java
Comment (TheMutsi) [reply packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java:655]: I implemented iOS first and they call it that way. should I change ? 
image

File: packages/spans/src/index.ts
Comment (TheMutsi) [reply packages/spans/src/index.ts:11]: JS and TS will pass it as undefined (similar to null)

File: packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java
Comment (TheMutsi) [reply packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java:669]: Note: We agreed to use Nanos until iOS change it

File: packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java
Comment (TheMutsi) [reply packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java:620]: I dont have Long in my "types whitelist"
image
2 lines bellow I cast it to long

File: packages/action-tracker/README.md
Comment (TheMutsi) [reply packages/action-tracker/README.md:31]: Yes we dont use it anymore

File: packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java
Comment (bidetofevil) [inline packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java:665]: Since this is open sourced, you might want to add a comment here saying the Android SDK expects millisecond but this works for now because of the workaround.

**Review by bidetofevil — APPROVED** [review]
Looks good from the Android side!

File: packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java
Comment (TheMutsi) [reply packages/core/android/src/main/java/io/embrace/embracewrapper/EmbraceManagerModule.java:665]: I will let it with that signature, since I need to match the parameters with iOS. In the next version , when iOS update the parameter, I will refactor this

**Pull request merged**
